### DISCUSSION TITLE
Proposal: return, break and continue expressions

### DISCUSSION THREAD

**Discussion by alrz** [dotnet/csharplang#176] [General]
Ported from https://github.com/dotnet/roslyn/issues/14239

# return, break and continue expressions

## Summary

Lets you use `return`, `break` and `continue` as an expression.

## Proposal

Since `throw` is now an expression, I think `return` , `break` and `continue` are also good candidates to be an expression. 

``` cs
var c = obj as C ?? return;
// instead of
if(!(obj is C c)) return;


var o = F() ?? return;
// instead of
var o = F();
if (o == null) return;

var (foo, bar) = GetFooBar() ?? return; // `GetFooBar` returns a nullable tuple
// instead of
if (!(GetFooBar() is var (foo, bar))) return;
// or
if (!TryGetFooBar(out var foo, out var bar)) return;
// or 
var result = GetFooBar();
if (result == null) return;
var (foo, bar) = result.Value;

foreach (var item in list) {
  var result = item match {
    case 1: "one",
    case 2: "two",
    case 3: break,
    default: continue,
  };
}
```

There is no straightforward alternative for the third example as you will need to use the `switch` statement.

**Open question:** `throw` expressions are not allowed as the RHS for `&&` and `||`. should that also be applied here?

~~Unlike `throw`,  these expressions need to return a value,~~
```cs
void F(out int i, out int j) {
  (i, j) = condition ? (1, 2) : return;
}
```
**EDIT:** We could make this specific example an error  as suggested by @lostmsu. 

**Open question:** should we allow `yield break` and `goto` as an expression? (`yield return` is excluded because it is not a jump instruction).

**Comment by Logerfo** [comment]
Combined with #102, we could get really objective code, like this one:
```c#
R M(object obj)
{
	var c = obj as C ?? return default;
	//...
}
```
Which would be equivalent to the code below with current language specs:
```c#
R M(object obj)
{
	if(obj is C c)
	{
		//...
	}
	//else not needed since we suppose there is a return inside the if block
	return default(R);
}
```
Which is far more objective than what we had using c# 6.0:
```c#
R M(object obj)
{
	var c = obj as C;
	if(c == null)
		return default(R);

	//...
}
```

**Comment by gulshan** [comment]
This feature will encourage mixing imperative and functional programming styles. That can turn out to be a bad idea IMHO.

**Comment by alrz** [comment]
Not sure how is this functional but mixing imperative and functional programming were never considered a bad idea and in fact it's one of the fundamental design directions that C# follows.

**Comment by gulshan** [comment]
This code contrasts among different coding styles-

```csharp
// Functional
var res1 = list
    .TakeWhile(n => n != -1)
    .Aggregate((i, j) => i + j);

// Imparetive
var res2 = 0;
foreach (var n in list)
{
    if (n == -1) break;
    res2 += n;
}

// Mixed, enabled by this proposal
var res3 = 0;
foreach (var n in list)
{
    res3 += (n != -1) ? n : break;
}
```

In functional style, everything is an expression. Where as in imperative style, everything is a statement. In functional languages loops are avoided and recursion is the way of repeating some task. While F# does allow loops, `break` and `continue` are not allowed. C# allows both styles. While mixing them is possible, not very intuitive now and people will gravitate towards either of the styles. But this proposal will enable more freestyle mix-up.

Newly added `throw` expression in C# corresponds to `failwith` expression in F#. And exceptions should be exceptional. I also want to quote [comment](https://github.com/dotnet/roslyn/issues/5143#issuecomment-159117188) by @gafter when this was first proposed within the discussion https://github.com/dotnet/roslyn/issues/5143 -
>`throw` is not a control statement in the same sense as `continue` and `break`, and they make less sense in an expression context.

**Comment by HaloFour** [comment]
One of the things that I like about the potential of using `continue` and `break` in expressions is the possibility of also extending LINQ query syntax to support them where they could translate into `Where` and `TakeWhile` respectively.

```cs
var query = from n in list
    select n ==  1 ? "one"
         : n ==  2 ? "two"
         : n == -1 ? break
         : continue;

// translates into

var query = list.Select(n =>
                  n ==  1 ? (true, true, "one")
                : n ==  2 ? (true, true, "two")
                : n == -1 ? (false, true, default(string))
                :           (true, false, default(string)))
    .TakeWhile(tuple => tuple.Item1)
    .Where(tuple => tuple.Item2)
    .Select(tuple => tuple.Item3);
```

As for the argument regarding imperative v. functional programming, most developers don't think in such black and white.  A functional feature in an imperative language isn't inherently a bad (or good) thing.  And while it's true that not everything is an expression in imperative languages, many things still are.  Beyond the idealistic argument, I can understand that it might be difficult to discern that an expression could affect the flow of the containing code if a control clause is buried deeply within.  But I don't think I'd doom the entire proposal based on what a bad/"clever" programmer can do to abuse it.

**Comment by Thaina** [comment]
Not oppose the idea just present what I normally did

```C#
void M(object obj)
{
    foreach (var item in list.Where((item) => item >= 1 && item <= 3).TakeWhile((it) => it != 3))
    {
        var result = item == 1 ? "one" : "two";
    }
}

// Actually I would like to explicit if break from the first line
void M(object obj)
{
    foreach (var item in list.Where((item) => item >= 1 && item <= 3))
    {
        if(item == 3)
            break;

        var result = item == 1 ? "one" : "two";
    }
}

```

Well, I think this feature has too many alternative

**Comment by lachbaer** [comment]
Proposal #135 could actually come in handy here, as well.

**Comment by alrz** [comment]
Since `return` would be allowed in a null-coalescing  operator,
```cs
foreach (var item in list ?? return)
```
I wonder if we could abuse `break` in that context to "skip" the loop instead of early return.
```cs
foreach (var item in list ?? break)
```

**Comment by lachbaer** [comment]
I don't see where
```C#
foreach (var item in list ?? return) {
...
}
```
~~would be more comfortable than putting a conditional return statement in the loop itself~~

```C#
// ** example is wrong ** just left for documentation
foreach (var item in list) {
    if (item == null) return;
...
}
```
~~Latter is even more intuitive to read. What binding does it actually have? To me it is a bit arbitrary.~~
Binding issue solved. But it should be written in braces, so that it is is clear to a not knowing reader.
```C#
// not this
(var item in list) ?? return
or
// it's this
var item in (list ?? return)
```

BTW, `goto` should also be added to the list for the sake of completeness. So all non-linear "jump" statements are respected.

**Comment by alrz** [comment]
@lachbaer 

>Latter is even more intuitive to read. 

Except that it will throw. You probably want to write it before the loop. 

Equivalent code of `?? break` would be actually a little messier.
```cs
if (list != null)
{
  foreach (var item in list)
  {
  }
}
```
Embedded statements can make this less verbose but I would certainly prefer `?? break` over that. 

>What binding does it actually have? 

`foreach` accepts an *expression* after `in`, so it is not arbitrary at all. In fact, you can already write this:
```cs
foreach (var item in list ?? throw new Exception(..))
```
But it's useless, because you are just converting a NRE to another exception.

> goto should also be added to the list for the sake of completeness.

I don't disagree, but I just didn't find any sensible use case for it. Personally, I try to not use `goto` beyond `goto default;` which is actually nice.

**Comment by benaadams** [comment]
Or
```csharp
foreach (var item in list ?? emptyList)
{
    // ...
}
```

**Comment by alrz** [comment]
@benaadams

If `list` is an array, yeah, that might be good enough, but with any other `IEnumerable`, you might seem a little apathetic to performance.

**Comment by lachbaer** [comment]
> I don't disagree, but I just didn't find any sensible use case for it. Personally, I try to not use goto beyond goto default; which is actually nice.

I don't use `goto` as well (except for debugging purposes), but despite liking or not it has its usage cases. But as it is a flow breaking statement, like `return`, `break`, `continue` and `throw`, this one should also be respected.

I like this proposal. C# has already come so far and there are already so many ways to reach the goal, including functional vs. imperative approaches, LINQ, wse. I think that there surely will be definite use cases for it where it will improve reading or the process of writing.

There are many that prefer not to jump out of a block in the middle! That also is probably just a matter of taste and personal experience. I would recommend to specially syntax highlight the occurences of these keywords in code editors, whether standing allone or being part of a longer expression!!!! (There are already extensions, but it should be standard).

**Comment by lostmsu** [comment]
Just wanted to note, that `return` expression is no-brainer: just like `throw` expression it always leaves current method, so there's no need to worry about what value does it have.

**Comment by alrz** [comment]
@lostmsu See the example in the OP with `out` parameters.

**Comment by lostmsu** [comment]
@alrz I don't think it is a valid use case. With the current execution order rules assignment should happen after `return` and thus should never be executed.

E.g. it should translate to (dropping one branch)
```csharp
if (!IsConstant(expression1))
  return false;
constant = expression1;
return true;
```
and so it should fail to compile due to unassigned `out` variable.

Doing otherwise would allow really weird constructs like
```csharp
object x, y;
y = Hello(x = someBool? blah: return foo);
```
Now it is clear, that return should prevent Hello from happening. And by the same logic it should stop `x =` and `y =` as well.

**Comment by lachbaer** [comment]
I proposed special syntax highlighting of branch keywords in dotnet/roslyn#18810. That would absolutely work in favor of this proposal.

Please give it a 👍 [over there](https://github.com/dotnet/roslyn/issues/18810#issue-222721543) if you think so, too.

**Comment by alrz** [comment]
I've added another example as an alternative to `Try` pattern that is extensively used in Roslyn,
```cs
// we can't deconstruct a nullable tuple but `??` unwraps it,
var (foo, bar) = GetFooBar() ?? return; // `GetFooBar` returns a nullable tuple
// instead of
if (!TryGetFooBar(out var foo, out var bar)) return; 
// ^ following Roslyn code style (using braces) would make this three lines longer!
```
Also, in `TryGetFooBar` we need to assign all `out` parameters even in case of failure. For `GetFooBar` we can just return `null` to indicate a failure.

**Comment by jnm2** [comment]
I'm starting to really want this feature. Code that looks like this:

```c#
var maybeFoo = TryGetFoo(...); // Returns Nullable<Foo>
if (maybeFoo == null) continue; // In other places, sometimes break or return. Def. not throw.

// Uses maybeFoo.Value a dozen times, annoying and hurts readability
DoSomething(maybeFoo.Value.Property1, maybeFoo.Value.Property2);
if (maybeFoo.Value.X == otherFoo.X && maybeFoo.Value.Y == otherFoo.Y && maybeFoo.Value.Z == otherFoo.Z)
{
    set.Remove(maybeFoo.Value);
}
```

Could look like this:

```c#
var foo = TryGetFoo(...) ?? continue;

DoSomething(foo.Property1, foo.Property2);
if (foo.X == otherFoo.X && foo.Y == otherFoo.Y && foo.Z == otherFoo.Z)
{
    set.Remove(foo);
}
```

**Comment by alrz** [comment]
A preliminary prototype is available over at: https://github.com/dotnet/roslyn/issues/20269

Note: `return` (without a trailing expression) is not supported yet due to a public API issue.

**Comment by alrz** [comment]
Thanks to @ashmind, a working build is deployed to [tryroslyn](https://sharplab.io/#v2:EYLghgNgTgXgtAMwKZgC4FcpIM5wFYC2ADnEgB5FQA+AAgAwAENAjANwCwAUDQMxMBMDAMIMA3lwaSmfAJYA7VAwCyACgD2wPEgDGqANoBdBmoCUYiVMsI1WMNoAWDFQDcwUBjNRICHucYYA/AEMwLYA1mailgwW0dFCanLYahBIAHQA6lCeSAAy8kgqOT5BDNqJqPLoSCYcnHGSAL6x0TQA7MrqgcHtDHS1Lc2cjUA=)
```cs
using System;
public class C {
    public int M(object[] o) {
        foreach (var item in o ?? break) {     
            Console.WriteLine(item ?? continue);
        }
        return M(o ?? return 0);
    }
}
```

**Comment by jnm2** [comment]
@alrz Why is the extra `DateTime?` local created?

https://sharplab.io/#v2:EYLghgNgTgXgtAMwKZgC4FcpIM5wFYC2ADnEgB5FQA+AAgAwAENAjANwCwAUFzQEwMBhLgG8GYhl3FMAzE2aMAsgAoAImiQAVAJYEkAfgYATddt0BKSeOGWpYgG5goDBxHRIGAXiMmd7vQZoAdgYAO3QICFZbcRtbIIYAEgAiYRc3ADoAQUNDNQBPbCVmMwBfABoGVMgM7MMFAHsQ1AALQuLyyrSkLJyATRQoNtKkjk4pEq4SoA=

**Comment by alrz** [comment]
It's being read in different places. what code you'd expect?

**Comment by jnm2** [comment]
@alrz Not the `DateTime` local, the `DateTime?` local.

Instead of:
```c#
DateTime? dateTime2 = dateTime;
if (!dateTime2.HasValue)
{
    return null;
}
DateTime valueOrDefault = dateTime2.GetValueOrDefault();
return string.Format("{0}, {1}, {2}", valueOrDefault.AddDays(1.0), valueOrDefault.AddMonths(1), valueOrDefault.AddYears(1));
```
I expected:

```c#
if (!dateTime.HasValue)
{
    return null;
}
DateTime valueOrDefault = dateTime.GetValueOrDefault();
return string.Format("{0}, {1}, {2}", valueOrDefault.AddDays(1.0), valueOrDefault.AddMonths(1), valueOrDefault.AddYears(1));
```

**Comment by jnm2** [comment]
Compare with a reference type: https://sharplab.io/#v2:EYLghgNgTgXgtAMwKZgC4FcpIM5wFYC2ADnEgB5FQA+AAgAwAENAjANwCwAUFzQEwMBhLgG8GYhl3FMAzE2aMAsgAoA9sDxIAxqgYATNEgAqASwJIAlJPHCrUsQDcwUBo4jokDALwMlAEQMmZuZK+qhGph4A/JFMAOwMAHboEBDmrHbitnY08QAkAETCru4AdACCurr+AJ7YSszmAL4ANAxFkKUVugoqCagAFnUNLW3FSOWVAJooUENN+RycUo1cjUA=

**Comment by HaloFour** [comment]
@jnm2 

The C# 7.0 compiler also copies struct arguments into locals when dealing with `throw` expressions, and probably in a lot of other scenarios.

**Comment by jnm2** [comment]
Oh, is this a defensive copy against `.HasValue` or `.GetValueOrDefault()` mutating the parameter? Seems like a known immutable value type like `Nullable<>` could be special-cased for performance.
It could also just be smart enough to realize that I'm not using the parameter any further down in the method.

**Comment by alrz** [comment]
@jnm2 I see. Appearantly that's how `??` works on nullables. Nothing special for return expressions.

**Comment by jnm2** [comment]
@alrz False alarm, sorry for sidetracking.

**Comment by jnm2** [comment]
@alrz Maybe it was the anticipation, but I'm enjoying this *way* too much 😆 We'd better be seeing this in the language!

**Comment by lostmsu** [comment]
@alrz , what is the public API issue?

**Comment by alrz** [comment]
@jnm2 

I hope so. But as it turns out, this does introduce some oddities into the language and underlying implementation. For example, we should not be able to use `continue` in the `foreach`'s expression, if we want to be able to use `break` there, we should separate things out into different binders. And that's just the beginning. Because in that case, `continue` will be bound to the outer loop (if any). I'm not sure if it could be the desirable behavior. Another thing is that if you use `break`, for example, inside a `return` statement, it should somehow cancel out the unreachable state.
```cs
switch (o) {
  case var x: return x as C ?? break;
}
return null; // false unreachable warning?
```
None of these are applied to `throw` expressions.

**Comment by alrz** [comment]
@lostmsu 

I asked around on [gitter](https://gitter.im/dotnet/roslyn?at=594acf23746f1fd66329590b), it might be a bug in the code generator or the public API analyzer. If you make a syntax optional, it generates optional parameters in the syntax factory. while we do have the exact same situation for `ReturnStatement`, it didn't work for `ReturnExpression`.

**Comment by HaloFour** [comment]
@alrz 

Conceptually I like the idea of using `break` within a `foreach` statement like that, however I think that introduces some unexpected inconsistency to the language.  The expression evaluated to produce the enumerable is always self-contained and behaves identically within a loop as well as not.  But that's not the case here.  What's worse is that depending on the context the expression itself would be valid, if it were to appear nested within another loop or maybe a `switch` case.

As such, given the following example:

```cs
var people = new Person[] {
    new Person() { Name = "Billy", Dependents = default(List<Person>) }
};
foreach (var person in people) {
    foreach (var dependent in person.Dependents ?? break) {
        Console.WriteLine($"{dependent.Name} is a dependent of {person.Name}.");
    }
    Console.WriteLine($"Finished listing the dependents of {person.Name}");
}
``` 

I'd expect the above to print nothing.

**Comment by HaloFour** [comment]
@alrz 

To put it another way, I'd expect the following two to behave in exactly the same manner:

```cs
// example 1
foreach (var foo in bar ?? break) { ... }

// example 2
var temp = bar ?? break;
foreach (var foo in temp) { ... }
```

**Comment by alrz** [comment]
@HaloFour 

Agreed, since we should correspond `continue`  to the outer loop anyways, it wouldn't be consistent to treat `break` differently. Without touching the binder, it will behave just like what you described.

**Comment by HaloFour** [comment]
@alrz 

Sounds good.  Kind of a shame that it couldn't be used to skip the current enumeration.  Maybe `goto` expressions?  Or if C# gets `break`/`continue` labels?

```cs
foreach (var foo in bar ?? goto label1) { ... }
label1:

label2: foreach (var foo in bar ?? break label2) { ... }
```

**Comment by jnm2** [comment]
@HaloFour what's the difference between a break label, a continue label, and a goto?

**Comment by HaloFour** [comment]
@jnm2 

Avoids the `goto` keyword, mostly, to which some people have a severe allergy.  Beyond that the difference really lies in the position of the label.  In Java, the label always goes immediately before the loop and using `break label` or `continue label` would either break out of or continue iteration of that loop from within a nested loop.

[Java Branching Statements](https://docs.oracle.com/javase/tutorial/java/nutsandbolts/branch.html)

**Comment by ashmind** [comment]
@HaloFour 
> Kind of a shame that it couldn't be used to skip the current enumeration.

There is always a possibility of future `foreach?` in the mold of #35, so this doesn't have to cover it.

**Comment by Logerfo** [comment]
@HaloFour I totally didn't know about Java Branching Statements. Is there something preventing C# to have this kind of features?
Also, I like the way [php](http://php.net/manual/pt_BR/control-structures.break.php) deals with this.

**Comment by HaloFour** [comment]
@Logerfo 

> Is there something preventing C# to have this kind of features?

Desire?  Philosophy?  C# has the feature, just with different syntax.  The only reason to suggest implementing Java's syntax is disliking having to use the keyword `goto`.  Not that I'm arguing against having a more structured form.

**Comment by Logerfo** [comment]
`goto` does work for the `break` case, but, as far as I know, not for the `continue` one.

**Comment by HaloFour** [comment]
@Logerfo 

You're right.  It can be accomplished through a combination of other things, such as `break` or `goto` out of a nested loop and checking a flag, but it's certainly not as succinct as a branched statement or labeled loop.

**Comment by AndreasHassing** [comment]
I'm carefully optimistic about this idea. I often find myself writing boilerplate code such as:
```csharp
var user = await userManager.GetUserAsync(User);
if (user == null) { return RedirectToLogin(); }
// or
var thing = await thingService.GetThingAsync(id);
if (thing == null) { return RedirectToIndex(); }
```

Which, with this proposal, could be:
```csharp
var user = await userManager.GetUserAsync(User) ?? return RedirectToLogin();
// or
var thing = await thingService.GetThingAsync(id) ?? return RedirectToIndex();
```

I think it reads quite nicely, and the verbosity from what I have now is not helping here.

**Comment by momvart** [comment]
Is this proposal going on in another thread? I wonder why it is inactive despite the fact that it is one of the most useful features.


After the addition of support for throwing exceptions in the null-coalescing operator, I was anticipating return statements to be added either. But after three years, it seems that it is not even one of the champions.

**Comment by OFark** [comment]
I have a [different take ](https://github.com/dotnet/csharplang/discussions/6345) on this proposal and swapped the syntax around. 

basically:
`return if thing is null : null`
or
`return if thing.IsInvalid : new InvalidResponse();`

so it's like the 'else if' statement

```
foreach(var item in items)
{
    continue if item.Value < 3;

    // do otherthings with Value >= 3;
}
```

**Reply by HaloFour** [reply]
Doing that limits where such an expression can be applied.  As an expression it can be applied anywhere an expression can be applied.

It also inverses the visibility of the important part of the statement, which is the condition.  Having to scan the entire line of every `return` all the way to the right to see if it might be conditional I feel makes the code substantially harder to read.  Whitespace is trivia in a language like C# and blocks are always optional.  If you are looking for the same degree of terseness, you can already do:

```cs
if (item.Value < 3) continue;
```

**Reply by OFark** [reply]
There's a group of devs that would say that if statement should have braces
```cs
if (item.Value < 3) 
{
   continue;
}
```

It was a conversation about this very thing that lead to my proposal for a condition for the return statement.

**Reply by OFark** [reply]
There is also the code smell of if statements inside if statements. The suggestion being break it out into another method, but you can't very well issue a return from inside the method called.

**Reply by HaloFour** [reply]
> There's a group of devs that would say that if statement should have braces

There's a group of devs that would say that `return if` and the like are confusing because they push the conditional off of the visible part of the screen.  In fact that is exactly why this proposal hasn't advanced.  The language can't invent a dozen different ways to do the same thing because each team has their own stylistic preferences.

> There is also the code smell of if statements inside if statements.

There's no code smell in nesting `if` statements or embedding control statements within conditionals.  That is idiomatic C#.

**Reply by Thaina** [reply]
There has been similar discussion in https://github.com/dotnet/csharplang/discussions/138#discussioncomment-65415

And also https://github.com/dotnet/roslyn/issues/11562

**Reply by HaloFour** [reply]
And: #2195

**Comment by davidnmbond** [comment]
This feature makes the code more succinct and its intentions are clear.  What's not to like?

**Comment by CyrusNajmabadi** [comment]
Folding labeled break/continue (https://github.com/dotnet/csharplang/discussions/6634) into a bucket with this may make improve the chances of these being done.

**Comment by noptools** [comment]
I agree.  Another example in a method with void return type:

var x = n switch
{
    1 => DoOne(),
    2 => DoTwo(),
    _ => return
}

(Should I start another 'discussion' to make this a 'proposal'?  Is there already one out there?)

**Reply by CyrusNajmabadi** [reply]
That would be covered under this proposal.

**Reply by DavidArno** [reply]
@noptools,

See #3038, #3037, #2632 and #2860. It's a popular request 😃

**Comment by jnm2** [comment]
If sequence expressions are implemented, then `?? { break; }` would have to work anyway. If that works, then why not allow the simplified form as well?

**Comment by Kuurama** [comment]
Would it allow syntax like that? (Assuming that's how default Result DU looks like):
```cs
var bar = await Foo.GetBarResultAsync() switch
{ 
    .Success(var bar) => bar,
    .Failure(var error) => return error;
};
```

Or simply with Nullable<Bar>?:
```cs
var bar = await Foo.GetBarNullableAsync() ?? return;
```

That could streamline quite a lot the usage of DUs!
Because this is less pretty in my opinion:
```cs
var barResult= await Foo.GetBarResultAsync();
if (!barResult.IsSuccess) 
   return barResult.Error;
            
var bar = barResult.Value;
            
// Or alternatively (But still a lot imperative and fairly wide to read):
if (!(await Foo.GetBarResultAsync()).TryGetValue(out var bar, out var error))
    return error;
```

**Reply by Kuurama** [reply]
My dream would maybe be something like this:
```cs
await Foo.GetBarResultAsync() ?? return .Error;
```

Where `??` could be applied to the future result type, and the `.` could grab the context just like the switch expression does.
So it would allow us to return the `res.Error`, but that would even be remotely possible only if this `return` as expression gets implemented.
Yeah, it would be close to Rust `?` operator to immediately return the error.
And since the C# type system could be tough on the error, we could even use this syntax to provide mapping with explicit casts or mapper after the `?? return`.

But obviously, that's another proposal. With the requirements being this current proposal, and default DUs.
But maybe insights like mine could give some ideas.

**Comment by jv-np** [comment]
I'm all for it, there are many cases a `nullable ?? return/continue/break` would be very nice, like in kotlin.
